### ISSUE TITLE
Difficult error in NuGet restore when Version env. var is set to a random string

### ISSUE THREAD

**Issue by KirillOsenkov** [NuGet/Home#6154]
## Details about Problem

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe):
NuGet in MSBuild with PackageReference (non-SDK style project)

NuGet version (x.x.x.xxx):
NuGet in Mono 5.8.0

OS version (i.e. win10 v1607 (14393.321)):
Mac OS but doesn't matter

## Detailed repro steps so we can see the same problem

1. Have a simple .proj file with a PackageReference 
2. Set the Version environment variable to, say, fbc5117dcb21f0d180c4b199d828d4f4b9d6f4a8
3. msbuild /t:Restore

Build fails with:
```
/Library/Frameworks/Mono.framework/Versions/5.8.0/lib/mono/xbuild/NuGet.targets(102,5): error : 'fbc5117dcb21f0d180c4b199d828d4f4b9d6f4a8' is not a valid version string. [/Users/builder/test/test.proj]
/Library/Frameworks/Mono.framework/Versions/5.8.0/lib/mono/xbuild/NuGet.targets(102,5): error : Parameter name: value
```

This is because of this logic:
https://github.com/NuGet/NuGet.Client/blob/2aad8d25e66e9c3737b220631f3661bb6235d14b/src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets#L555-L565

If the environment defines a variable called `Version` or `PackageVersion` and it can't be parsed as a valid SemanticVersion, the build fails with the inscrutable error above. 

Call stack:

```
System.ArgumentException: 'fbc5117dcb21f0d180c4b199d828d4f4b9d6f4a8' is not a valid version string.
Parameter name: value
  at NuGet.Versioning.NuGetVersion.Parse (System.String value) [0x00051] in <e60c4bcdac3f4fa4aaa3750db25cd7a6>:0 
  at NuGet.Commands.MSBuildRestoreUtility.GetVersion (NuGet.Commands.IMSBuildItem item) [0x00021] in <2e3b454e9a2d45eb90859e27e70c3278>:0 
  at NuGet.Commands.MSBuildRestoreUtility.GetPackageSpec (System.Collections.Generic.IEnumerable`1[T] items) [0x00256] in <2e3b454e9a2d45eb90859e27e70c3278>:0 
  at System.Linq.Enumerable+SelectEnumerableIterator`2[TSource,TResult].MoveNext () [0x00048] in <f7cad77794fc4fa5bf9a33531f4b94b4>:0 
  at System.Linq.Enumerable+WhereEnumerableIterator`1[TSource].MoveNext () [0x0004e] in <f7cad77794fc4fa5bf9a33531f4b94b4>:0 
  at NuGet.Commands.MSBuildRestoreUtility.GetDependencySpec (System.Collections.Generic.IEnumerable`1[T] items) [0x0017e] in <2e3b454e9a2d45eb90859e27e70c3278>:0 
  at NuGet.Build.Tasks.RestoreTask+<ExecuteAsync>d__31.MoveNext () [0x00092] in <696d47d17c774b4caf18505e4b7b9cff>:0 
```

I doubt this can be fixed, but I imagine it would be possible to add an early validation to the RestoreTask here:
https://github.com/NuGet/NuGet.Client/blob/2aad8d25e66e9c3737b220631f3661bb6235d14b/src/NuGet.Core/NuGet.Build.Tasks/RestoreTask.cs#L84

Basically look at the Version of RestoreGraphItems and if you TryParse it as a semantic version and it fails, issue an early error message like: 
```
Version of RestoreGraphItems <GUID> of type <ProjectSpec> can't be parsed as a valid semantic version. The value of the Version property is: $(Version) The value of the PackageVersion is: $(PackageVersion)
```

Contact kirillo internally for a .binlog file and all the repro details.

**Comment by emgarten** [comment]
@KirillOsenkov thanks for the issue. I think this is best tracked here: https://github.com/NuGet/Home/issues/3601

It is a little different for env vars, but as far as I know msbuild doesn't provide a good way to tell the difference so that we can tell the user where the bad data came from.

**Comment by crbreingan** [comment]
I just ran into this issue too. I had no idea that MSBuild or nuget was going to use the version environment variable. Is this documented somewhere? It conflicts with my build server pipelines, so now I have to figure out how to change everything else because nuget is using a super common word from the environment.
